### PR TITLE
💄 :new: Functions Abstraction Rework

### PR DESCRIPTION
**Pull request by jameshughes89**
### Related Issues or PRs
Noted in #137 

### What
Add more content to the abstraction section. Break the section down into subsections. Add a warning emphasizing that it is non-trivial. 

### Additional Notes
Let me know if the warning is *tooooo* much.

### CODE REVIEW

File: site/topics/functions/functions.rst [hunk 1]
```diff
@@ -226,15 +226,41 @@ Abstraction
...
+* Knowing how and where to create levels of abstraction requires a deep understanding of the problem you are trying to address
+
+.. warning::
+
+    It needs to be emphasized that how, when, and where to create levels of abstraction can be very difficult. This is
+    not something you can learn by just reading about it; this is a skill that you will be developing throughout your
+    careers. Over time you will get better and better at it.
+
+    Often there isn't even a *right* way to do it. Some may be better than others, but that does not mean that the
+    others are not "correct".
```
Comment (twentylemon) [inline site/topics/functions/functions.rst:260]: Minor, it'd be "correct." Quotes go after punctuation, for whatever reason.

**Review by twentylemon — APPROVED** [review]
Seems reasonable enough. Maybe `warning` -> `note` or something would be the only thing I'd suggest to reduce the scare factor.

File: site/topics/functions/functions.rst (hunk 1 above)
Comment (jameshughes89) [reply site/topics/functions/functions.rst:260]: I reject the correctness of "correct." for the same reasons you feel.

Comment (twentylemon) [reply site/topics/functions/functions.rst:260]: That one should probably be "correct," bruh.

Comment (jameshughes89) [reply site/topics/functions/functions.rst:260]: ew

**Pull request merged**